File: app/src/main/java/com/huanchengfly/tieba/post/widgets/theme/TintTextView.java
package com.huanchengfly.tieba.post.widgets.theme;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.appcompat.widget.AppCompatTextView;

import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.ui.theme.interfaces.Tintable;
import com.huanchengfly.tieba.post.ui.theme.utils.ColorStateListUtils;
import com.huanchengfly.tieba.post.ui.theme.utils.ThemeUtils;

@SuppressLint("CustomViewStyleable")
public class TintTextView extends AppCompatTextView implements Tintable {
    private int mBackgroundTintResId;
    private int mTintResId;

    public TintTextView(Context context) {
        this(context, null);
    }

    public TintTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TintTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        if (isInEditMode()) {
            return;
        }
        if (attrs == null) {
            mBackgroundTintResId = 0;
            mTintResId = 0;
            applyTintColor();
            return;
        }
        TypedArray array = getContext().obtainStyledAttributes(attrs, R.styleable.TintView, defStyleAttr, 0);
        mBackgroundTintResId = array.getResourceId(R.styleable.TintView_backgroundTint, 0);
        mTintResId = array.getResourceId(R.styleable.TintView_tint, 0);
        array.recycle();
        applyTintColor();
    }

    public void setBackgroundTintResId(int mBackgroundTintResId) {
        this.mBackgroundTintResId = mBackgroundTintResId;
        applyTintColor();
    }

    public void setTintResId(int mTintResId) {
        this.mTintResId = mTintResId;
        applyTintColor();
    }

    private void applyTintColor() {
        if (mBackgroundTintResId != 0) {
            if (getBackground() == null) {
                setBackgroundColor(ThemeUtils.getColorById(getContext(), mBackgroundTintResId));
            } else {
                setBackgroundTintList(ColorStateListUtils.createColorStateList(getContext(), mBackgroundTintResId));
            }
        }
        if (mTintResId != 0) {
            setTextColor(ColorStateListUtils.createColorStateList(getContext(), mTintResId));
        }
    }

    @Override
    public void tint() {
        applyTintColor();
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/widgets/theme/TintTextView.kt
package com.huanchengfly.tieba.post.widgets.theme

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.getColorStateListCompat
import com.huanchengfly.tieba.post.ui.theme.interfaces.Tintable
import com.huanchengfly.tieba.post.ui.theme.utils.ColorStateListUtils

@SuppressLint("CustomViewStyleable")
class TintTextView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatTextView(context!!, attrs, defStyleAttr), Tintable {
    var backgroundTintResId: Int = 0
        set(value) {
            field = value
            applyTintColor()
        }
    var tintResId: Int = 0
        set(value) {
            field = value
            applyTintColor()
        }

    private fun applyTintColor() {
        if (backgroundTintResId != 0) {
            if (background == null) {
                background = ColorDrawable(Color.WHITE)
            }
            backgroundTintList = if (isInEditMode) {
                context.getColorStateListCompat(backgroundTintResId)
            } else {
                ColorStateListUtils.createColorStateList(context, backgroundTintResId)
            }
        }
        if (tintResId != 0) {
            setTextColor(
                if (isInEditMode) {
                    context.getColorStateListCompat(tintResId)
                } else {
                    ColorStateListUtils.createColorStateList(context, tintResId)
                }
            )
        }
    }


    override fun tint() {
        applyTintColor()
    }

    init {
        val array =
            getContext().obtainStyledAttributes(attrs, R.styleable.TintView, defStyleAttr, 0)
        backgroundTintResId = array.getResourceId(R.styleable.TintView_backgroundTint, 0)
        tintResId = array.getResourceId(R.styleable.TintView_tint, 0)
        array.recycle()
    }
}